File: 1028._Recover_a_Tree_From_Preorder_Traversal.cpp
// https://leetcode.com/problems/recover-a-tree-from-preorder-traversal/?envType=daily-question&envId=2025-02-22
// Taken from:  https://www.youtube.com/watch?v=WN0quIK89F0&ab_channel=ChaiaurCode
// This is a HARD leveled question. Please revise this problem

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution
{
    int n;

    TreeNode *solve(string &s, int &i, int depth)
    {
        if (i >= n)
            return NULL;

        int j = i;
        while (j < n && s[j] == '-')
        {
            j++;
        }
        int dash = j - i;
        if (dash != depth)
        {
            return NULL;
        }

        // i =  j;
        i += dash;

        int num = 0;
        while (i < n && isdigit(s[i]))
        {
            num = (num * 10) + (s[i] - '0');
            i++;
        }

        TreeNode *root = new TreeNode(num);
        root->left = solve(s, i, depth + 1);
        root->right = solve(s, i, depth + 1);

        return root;
    }

public:
    TreeNode *recoverFromPreorder(string s)
    {
        n = s.length();
        int i = 0;
        int depth = 0;

        return solve(s, i, depth);
    }
};

File: 1028._Recover_a_Tree_From_Preorder_Traversal.java
// https://leetcode.com/problems/recover-a-tree-from-preorder-traversal/?envType=daily-question&envId=2025-02-22
// https://www.youtube.com/watch?v=ZhO5UPsasEM&t=114s&ab_channel=leetcode
// Revise this problem. 

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public TreeNode recoverFromPreorder(String S) {
        int level;
        int val;
        Stack<TreeNode> stack = new Stack<>();
        for (int i = 0; i < S.length();) {
            for (level = 0; S.charAt(i) == '-'; i++) {
                level++;
            }

            for (val = 0; i < S.length() && S.charAt(i) != '-'; i++) {
                val = val * 10 + (S.charAt(i) - '0');
            }

            while (stack.size() > level)
                stack.pop();

            TreeNode node = new TreeNode(val);

            if (!stack.isEmpty()) {
                if (stack.peek().left == null) {
                    stack.peek().left = node;
                } else {
                    stack.peek().right = node;
                }
            }
            stack.add(node);

        }

        while (stack.size() > 1)
            stack.pop();

        return stack.pop();
    }
}